### PR TITLE
Add query to match unchanged beliefs

### PR DESCRIPTION
**Pull request by Flix6x**
- [x] Add query util to match unchanged beliefs.
- [x] Test the new util function.
- [x] Add automated tests.
- [x] ~Move filters from `beliefs/classes.py` to `beliefs/queries.py`.~

### CODE REVIEW

**Comment by Flix6x** [comment]
Posting some test results here for discussion.

The belief history for a single event, **before** removing unchanged beliefs:
```
      event_start       |  belief_horizon   | cumulative_probability | event_value | sensor_id | source_id 
------------------------+-------------------+------------------------+-------------+-----------+-----------
 2021-05-16 21:00:00+00 | -1 days +23:55:00 |                    0.5 |         1.8 |        10 |         8
 2021-05-16 21:00:00+00 | 00:55:00          |                    0.5 |         1.8 |        10 |         8
 2021-05-16 21:00:00+00 | 01:56:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 02:55:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 03:56:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 04:56:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 05:55:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 06:55:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 07:55:00          |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 08:55:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 09:55:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 10:55:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 11:55:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 12:55:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 13:54:00          |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 14:54:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 15:54:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 16:54:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 17:55:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 18:55:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 19:54:00          |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 20:54:00          |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 21:55:00          |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 22:55:00          |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 23:55:00          |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 00:55:00    |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 01:55:00    |                    0.5 |        2.03 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 02:55:00    |                    0.5 |        2.03 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 03:55:00    |                    0.5 |        2.03 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 04:55:00    |                    0.5 |        2.03 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 05:55:00    |                    0.5 |        2.03 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 06:25:00    |                    0.5 |        2.03 |        10 |         8
(32 rows)
```

**After** removing unchanged beliefs:

```
      event_start       | belief_horizon | cumulative_probability | event_value | sensor_id | source_id 
------------------------+----------------+------------------------+-------------+-----------+-----------
 2021-05-16 21:00:00+00 | 00:55:00       |                    0.5 |         1.8 |        10 |         8
 2021-05-16 21:00:00+00 | 07:55:00       |                    0.5 |        2.04 |        10 |         8
 2021-05-16 21:00:00+00 | 13:54:00       |                    0.5 |        1.88 |        10 |         8
 2021-05-16 21:00:00+00 | 19:54:00       |                    0.5 |        1.85 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 00:55:00 |                    0.5 |        1.97 |        10 |         8
 2021-05-16 21:00:00+00 | 1 day 06:25:00 |                    0.5 |        2.03 |        10 |         8
(6 rows)
```

So far, I have only one issue with this: measurements (i.e. ex-post beliefs, which have a negative horizon) get removed in case their most recent forecast (smallest positive horizon) was exactly accurate. I don't like that, because even if that value doesn't represent a changed belief, the passing of knowledge time represent a significant state change. I have to think about this a bit, but I should probably be able to add an option to keep the first ex-post belief.

**Comment by nhoening** [comment]
Interesting the more one thinks about this... even without passing the positive/negative horizon boundary.

If I had a forecast with horizon 6 hours. Then I make another one with horizon 3 hours, which turns out to have the same value.

Which one do I want to keep? We now keep the older one, but some scheduler might be built which prefers forecasts where the horizon is small (if there is a choice, e.g. if there are multiple sources)

**Comment by Flix6x** [comment]
> [...] some scheduler might be built which prefers forecasts where the horizon is small (if there is a choice, e.g. if there are multiple sources)

I think that preference carries the implicit notion that smaller horizons would denote more certainty. If that is actually the case, the forecaster can make it explicit by saving probabilistic forecasts, and assigning more certainty to the belief with the smaller horizon.

**Comment by Flix6x** [comment]
I found a second issue. Here's another case that the current query doesn't catch yet. Notice the value of 13.07? That was the forecast 1 day 19:55:00 hours before knowledge time, after which it changed, but then 07:54:00 hours before knowledge time it became the forecast again.

```
      event_start       |  belief_horizon   | cumulative_probability | event_value | sensor_id | source_id 
------------------------+-------------------+------------------------+-------------+-----------+-----------
 2021-05-17 16:00:00+00 | -1 days +23:54:00 |                    0.5 |       14.27 |         9 |         8
 2021-05-17 16:00:00+00 | 00:54:00          |                    0.5 |       14.17 |         9 |         8
 2021-05-17 16:00:00+00 | 01:54:00          |                    0.5 |        13.7 |         9 |         8
 2021-05-17 16:00:00+00 | 02:55:00          |                    0.5 |       13.42 |         9 |         8
 2021-05-17 16:00:00+00 | 03:55:00          |                    0.5 |       12.99 |         9 |         8
 2021-05-17 16:00:00+00 | 04:55:00          |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 05:54:00          |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 06:55:00          |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 07:54:00          |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 08:55:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 09:54:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 10:54:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 11:54:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 12:55:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 13:55:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 14:54:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 15:53:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 16:55:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 17:55:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 18:55:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 19:55:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 20:56:00          |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 21:55:00          |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 22:56:00          |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 23:56:00          |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 00:55:00    |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 01:55:00    |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 02:55:00    |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 03:55:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 04:55:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 05:55:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 06:55:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 07:55:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 08:54:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 09:54:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 10:54:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 11:54:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 12:55:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 13:55:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 14:54:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 15:54:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 16:55:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 17:55:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 18:55:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 19:55:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 20:55:00    |                    0.5 |       13.09 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 21:55:00    |                    0.5 |       13.09 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 22:55:00    |                    0.5 |       13.09 |         9 |         8
(48 rows)
```

We want this to become
```
      event_start       |  belief_horizon   | cumulative_probability | event_value | sensor_id | source_id 
------------------------+-------------------+------------------------+-------------+-----------+-----------
 2021-05-17 16:00:00+00 | -1 days +23:54:00 |                    0.5 |       14.27 |         9 |         8
 2021-05-17 16:00:00+00 | 00:54:00          |                    0.5 |       14.17 |         9 |         8
 2021-05-17 16:00:00+00 | 01:54:00          |                    0.5 |        13.7 |         9 |         8
 2021-05-17 16:00:00+00 | 02:55:00          |                    0.5 |       13.42 |         9 |         8
 2021-05-17 16:00:00+00 | 03:55:00          |                    0.5 |       12.99 |         9 |         8
 2021-05-17 16:00:00+00 | 07:54:00          |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 13:55:00          |                    0.5 |       13.17 |         9 |         8
 2021-05-17 16:00:00+00 | 19:55:00          |                    0.5 |       13.08 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 02:55:00    |                    0.5 |       13.98 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 08:54:00    |                    0.5 |       13.12 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 14:54:00    |                    0.5 |       13.95 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 19:55:00    |                    0.5 |       13.07 |         9 |         8
 2021-05-17 16:00:00+00 | 1 day 22:55:00    |                    0.5 |       13.09 |         9 |         8
(13 rows)
```
However, the current query drops the row with the horizon of 07:54:00 hours, because that value also exists with a higher belief horizon. I have already written the raw SQL query that would resolve this issue (based on [this Stack Exchange issue](https://dba.stackexchange.com/questions/82820/how-can-i-remove-only-consecutive-duplicate-fields-when-the-id-is-not-consecutiv/82850#82850?newreg=b710cae7e728474fadd5cb9433605c53)), but I still need to get it into SQLAlchemy syntax.

File: timely_beliefs/beliefs/queries.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                tb3.belief_horizon > tb1.belief_horizon,
+                tb3.event_start == tb1.event_start,
+                tb3.sensor_id == tb1.sensor_id,
+                tb3.source_id == tb1.source_id,
+                *tb3_criteria,
+            )
+            .scalar_subquery(),
+        ),
+    ).where(
+        or_(tb2.source_id == tb1.source_id, tb2.source_id is None),
```
Comment (nhoening) [inline timely_beliefs/beliefs/queries.py:55]: Why this again? It's also in line 41

**Review by nhoening — CHANGES_REQUESTED** [review]
Looks nice.

I would like to ask for a test or two. Not sure how cumbersome they are to make, but you do provide good examples in the discussion of this PR.

File: timely_beliefs/beliefs/queries.py (hunk 1 above)
Comment (Flix6x) [reply timely_beliefs/beliefs/queries.py:55]: Thanks. This made me realize I could simplify a bit further.

Historical note: I had added this line to support the opposite query (getting changed beliefs), in which the case the very oldest belief in tb1 has all None values in tb2.

**Review by nhoening — APPROVED** [review]

**Pull request merged**